pkg/storage-init/storage-init.sh

```diff
@@ -34,6 +34,57 @@ fi
 
 P3_FS_TYPE="ext3"
 FSCK_FAILED=0
+
+# First lets see if we're missing P3 altogether and try to create it.
+# This is safe, since the worst case scenario we may hose the system,
+# but the system without P3 is a warm brick anyways
+P3=$(/hostfs/sbin/findfs PARTLABEL=P3)
+IMGA=$(/hostfs/sbin/findfs PARTLABEL=IMGA)
+if [ -z "$P3" ] && [ -n "$IMGA" ]; then
+   DEV=$(echo /sys/block/*/"${IMGA#/dev/}")
+   DEV="/dev/$(echo "$DEV" | cut -f4 -d/)"
+
+   # if sgdisk complains we need to repair the GPT
+   if sgdisk -v "$DEV" | grep -q 'Identified.*problems'; then
+       # save a copy of the MBR + first partition entry
+       # the logic here is that whatever booted us was good
+       # enough to get us here, so we'd rather sgdisk disn't
+       # mess up a good thing for us
+       dd if="$DEV" of=/tmp/mbr.bin bs=1 count=$((446 + 16)) conv=noerror,sync,notrunc
+
+       sgdisk -h1 -e "$DEV"
+
+       # move 1st MBR entry to 2nd place
+       dd if="$DEV" of="$DEV" bs=1 skip=446 seek=$(( 446 + 16)) count=16 conv=noerror,sync,notrunc
+       # restore 1st MBR entry + first partition entry
+       dd if=/tmp/mbr.bin of="$DEV" bs=1 conv=noerror,sync,notrunc
+
+       # focrce kernel to re-scan partition table
+       partprobe "$DEV"
+   fi
+
+   # lets see if IMGB partition is around, if not - create it
+   IMGB=$(/hostfs/sbin/findfs PARTLABEL=IMGB)
```

Comment:
Can we move this up to [here](https://github.com/lf-edge/eve/pull/870/files#diff-043db85bf353990b42dd8c985c767eddR42), so we get all of them at once, easier to read?

Reply:
Sure. Done @deitch